pkg/cloud/batch/batch.go

```diff
@@ -19,46 +19,178 @@ package batch
+}
+
+func (l *LocalBatchService) GetState() State {
+	return maps.Clone(l.state)
+}
+
+func (l *LocalBatchService) publishAction(action ActionState) {
+	l.bus.Publish(localBatchDeliveryTopic, action)
+}
+
+func (l *LocalBatchService) SubscribeToAction(subscription func(ActionState)) {
+	// ignore the error, it's only returned if the fn param isn't a function
+	_ = l.bus.Subscribe(localBatchDeliveryTopic, subscription)
+}
+
+func (l *LocalBatchService) registerJob(serviceName string, registration *batchpb.RegistrationRequest) {
+	l.batchLock.Lock()
+	defer l.batchLock.Unlock()
+
+	if l.state[registration.JobName] == nil {
+		l.state[registration.JobName] = make(map[string]int)
+	}
+
+	l.state[registration.JobName][serviceName]++
+
+	l.publishState()
+}
+
+func (l *LocalBatchService) unregisterJob(serviceName string, registration *batchpb.RegistrationRequest) {
+	l.batchLock.Lock()
+	defer l.batchLock.Unlock()
+
+	if l.state[registration.JobName] == nil {
+		l.state[registration.JobName] = make(map[string]int)
+	}
+
+	l.state[registration.JobName][serviceName]--
+
+	if l.state[registration.JobName][serviceName] == 0 {
+		delete(l.state, registration.JobName)
+	}
+
+	l.publishState()
+}
+
+func (l *LocalBatchService) HandleJob(stream batchpb.Job_HandleJobServer) error {
+	serviceName, err := grpcx.GetServiceNameFromStream(stream)
+	if err != nil {
+		return err
+	}
+
+	peekableStream := grpcx.NewPeekableStreamServer[*batchpb.ServerMessage, *batchpb.ClientMessage](stream)
+
+	firstRequest, err := peekableStream.Peek()
+	if err != nil {
+		return err
+	}
+
+	if firstRequest.GetRegistrationRequest() == nil {
+		return fmt.Errorf("first request must be a registration request")
+	}
+
+	err = stream.Send(&batchpb.ServerMessage{
+		Id: firstRequest.Id,
+		Content: &batchpb.ServerMessage_RegistrationResponse{
+			RegistrationResponse: &batchpb.RegistrationResponse{},
+		},
+	})
+	if err != nil {
+		return err
+	}
+
+	// Keep track of our local batch subscriptions
+	l.registerJob(serviceName, firstRequest.GetRegistrationRequest())
+	defer l.unregisterJob(serviceName, firstRequest.GetRegistrationRequest())
+
+	return l.JobManager.HandleJob(peekableStream)
+}
+
 func (l *LocalBatchService) SubmitJob(ctx context.Context, req *batchpb.JobSubmitRequest) (*batchpb.JobSubmitResponse, error) {
-	// TODO: Error if job does not exist
-	// Execute the job request in the background
-	go func() {
-		_, err := l.HandleJobRequest(&batchpb.ServerMessage{
-			Content: &batchpb.ServerMessage_JobRequest{
-				JobRequest: &batchpb.JobRequest{
-					JobName: req.GetJobName(),
-					Data:    req.Data,
-				},
+	json, err := req.Data.GetStruct().MarshalJSON()
+	if err != nil {
+		fmt.Println("Error marshalling job request data: ", err)
+		return nil, err
+	}
+
+	_, err = l.HandleJobRequest(&batchpb.ServerMessage{
+		Content: &batchpb.ServerMessage_JobRequest{
+			JobRequest: &batchpb.JobRequest{
+				JobName: req.GetJobName(),
+				Data:    req.Data,
 			},
+		},
+	})
```

Comment:
Job runs should be asynchronous, this change makes them synchronous. If you have a job that runs for 30mins this will block the return of the calling endpoint.